[Curl] Add clearing network cache for curl port

kshukuwa opened the pull request:
#### 990956f8884c9c341fc2b978d5aa5db005edd173
<pre>
[Curl] Add clearing network cache for curl port
<a href="https://bugs.webkit.org/show_bug.cgi?id=247355">https://bugs.webkit.org/show_bug.cgi?id=247355</a>

Reviewed by Alex Christensen.

Add clearing network cache for curl port. Also, add an C API
to clear network caches.

* Source/WebKit/NetworkProcess/curl/NetworkProcessCurl.cpp:
(WebKit::NetworkProcess::clearDiskCache):
* Source/WebKit/UIProcess/API/C/WKWebsiteDataStoreRef.cpp:
(WKWebsiteDataStoreRemoveNetworkCache):
* Source/WebKit/UIProcess/API/C/WKWebsiteDataStoreRef.h:
* Tools/MiniBrowser/win/WebKitBrowserWindow.cpp:
(removeAllFetchCachesCallback):
(removeNetworkCacheCallback):
(removeMemoryCachesCallback):
(removeAllServiceWorkerRegistrationsCallback):
(WebKitBrowserWindow::clearWebsiteData):
(removeAllFetchCaches): Deleted.
(removeMemoryCaches): Deleted.
(removeAllServiceWorkerRegistrations): Deleted.

Canonical link: <a href="https://commits.webkit.org/256437@main">https://commits.webkit.org/256437@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8a9a5ed7067872657fe5d077521e97eb0b9dd4bc

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/95648 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/4909 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/28696 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/105226 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/165526 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/99633 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/4975 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/33664 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/88022 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/101075 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/101309 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/3651 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/82266 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/30706 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/85519 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/87429 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/73540 "Found 1 new API test failure: /WebKitGTK/TestWebKitAccessibility:/webkit/WebKitAccessibility/accessible/event-listener (failure)") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/39398 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/18973 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/37097 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/20283 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4450 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/41093 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/42932 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/43080 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/39534 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebKit/UIProcess/API/C/WKWebsiteDataStoreRef.cpp:
```diff
@@ -757,6 +757,14 @@ void WKWebsiteDataStoreRemoveAllFetchCaches(WKWebsiteDataStoreRef dataStoreRef,
     });
 }
 
+void WKWebsiteDataStoreRemoveNetworkCache(WKWebsiteDataStoreRef dataStoreRef, void* context, WKWebsiteDataStoreRemoveNetworkCacheCallback callback)
```
I think this should have "DiskCache" in its name instead of "NetworkCache"
Even better would be to make a C API more like - (void)removeDataOfTypes:(NSSet<NSString *> *)dataTypes modifiedSince:(NSDate *)date completionHandler:(void (^)(void))completionHandler; where you can specify what types to remove rather than needing to make separate calls for each types.

kshukuwa replied:
> I think this should have "DiskCache" in its name instead of "NetworkCache"

I put NetworkCache in this function, referring to the function below. I think it is better to use the same name.

https://github.com/WebKit/WebKit/blob/main/Source/WebKit/UIProcess/API/C/WKWebsiteDataStoreConfigurationRef.h#L42-L43
```
WK_EXPORT WKStringRef WKWebsiteDataStoreConfigurationCopyNetworkCacheDirectory(WKWebsiteDataStoreConfigurationRef configuration);
WK_EXPORT void WKWebsiteDataStoreConfigurationSetNetworkCacheDirectory(WKWebsiteDataStoreConfigurationRef configuration, WKStringRef directory);
```

kshukuwa replied:
> Even better would be to make a C API more like ...

I see. This is simpler and more extensible. I'll try to implement the API.

achristensen07 replied:
Hmm.  You're right.  This does follow the design of the WKWebsiteDataStoreRef API.  It would still be nice to have an API to clear all the types.  Much more future-proof and fewer roundtrips.

kshukuwa committed:
[Curl] Add clearing network cache for curl port
https://bugs.webkit.org/show_bug.cgi?id=247355

Reviewed by Alex Christensen.

Add clearing network cache for curl port. Also, add an C API
to clear network caches.

* Source/WebKit/NetworkProcess/curl/NetworkProcessCurl.cpp:
(WebKit::NetworkProcess::clearDiskCache):
* Source/WebKit/UIProcess/API/C/WKWebsiteDataStoreRef.cpp:
(WKWebsiteDataStoreRemoveNetworkCache):
* Source/WebKit/UIProcess/API/C/WKWebsiteDataStoreRef.h:
* Tools/MiniBrowser/win/WebKitBrowserWindow.cpp:
(removeAllFetchCachesCallback):
(removeNetworkCacheCallback):
(removeMemoryCachesCallback):
(removeAllServiceWorkerRegistrationsCallback):
(WebKitBrowserWindow::clearWebsiteData):
(removeAllFetchCaches): Deleted.
(removeMemoryCaches): Deleted.
(removeAllServiceWorkerRegistrations): Deleted.

Canonical link: https://commits.webkit.org/256437@main
